### DISCUSSION TITLE
Be able to Deserialize a ActiveModel directly

### DISCUSSION THREAD

**Discussion by hf29h8sh321** [SeaQL/sea-orm#603] [General]
## Summary

Be able to Deserialize a ActiveModel directly

## Motivation

Insert a model without certain columns (such as no primary key, nothing in nullable column)

**Comment by billy1624** [comment]
Hey @hf29h8sh321, welcome! This has been discussed before. Let me know your thoughts on it!
- #457

**Comment by hf29h8sh321** [comment]
@billy1624 I looked into that discussion, but my motivation is different. If the HTTP Client wants to create a new task (example in the discussion), they only provide title and text, and not provide any id. The only way right now would be to create a separate struct that does not contain the id, and map each key to the ActiveModel.

**Comment by billy1624** [comment]
I see, perhaps this will be helpful?
- #346
- #492

**Comment by hf29h8sh321** [comment]
I tested #492 and it does not accept json values that do not have certain columns. 

Like the fruit example,

```rs
 #[derive(Clone, Debug, PartialEq, DeriveEntityModel)] 
 #[cfg_attr(feature = "with-json", derive(Serialize, Deserialize))] 
 #[sea_orm(table_name = "fruit")] 
 pub struct Model { 
     #[sea_orm(primary_key)] 
     #[serde(skip_deserializing)] 
     pub id: i32, 
     pub name: String, 
     pub cake_id: Option<i32>, 
 } 
```

If the json value is written to change `name` to something different, json is only `{ "name": "New Name" }` and is missing `cake_id`. It should be possible to convert the partial json data into ActiveModel with only one `Set`.

**Comment by tyt2y3** [comment]
The problem is two stage now: Deserialize a json and converting it into an ActiveModel
I think the newly added API should have solved the problem

**Reply by hf29h8sh321** [reply]
What would json be deserialized as ? Deserializing as model would require every field to be present

**Reply by billy1624** [reply]
Hey @hf29h8sh321, sorry for the delay.

The `set_from_json` method is operate on `ActiveModel` instead of `Model`.

> What would json be deserialized as ? Deserializing as model would require every field to be present

So, I guess this is not a problem? Please correct me if I'm wrong.

https://github.com/SeaQL/sea-orm/blob/ea5b0b1073f1ea6484a3a3471b3cd8e13f8d83eb/src/entity/active_model.rs#L938-L950

**Comment by jimmycuadra** [comment]
I have this same problem and have had a hard time figuring out how to do this based on the documentation and examples. It seems like it'd be an extremely common use case. I think `ActiveModelBehavior::set_from_json` is _close_ to right, but it only accepts a `serde_json::Value`, which means it could be arbitrary JSON. I think what would be needed is an equivalent of this function that takes any type that implements `serde::Deserialize` so I can control exactly which fields are allowed to be set for a given HTTP API.

**Comment by tyt2y3** [comment]
https://github.com/SeaQL/sea-orm/blob/2f4c4812803bc5c5c79259ff2f7edf32e2df52c4/examples/basic/src/operation.rs#L69-L77

You can create a custom ActiveModel and also derive Deserialize on it.

**Reply by jimmycuadra** [reply]
I knew about that technique and was trying to use it, which is how I ran into #1041. The missing `time` support aside, a custom ActiveModel still didn't work for my use case, but I think it's a different issue than what this discussion is about, so I'll probably just start a separate discussion.

**Reply by billy1624** [reply]
Hey @jimmycuadra, if you need a strict control to map user input to ActiveModel. I'd suggest you to define a struct on your own to contain the user input from for example a REST API, then `impl From<UserInput> for ActiveModel`.

But yeah, maybe we should start a new thread to discuss this. Would you mind submitting a feature request (issue) for this?
